ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,284 @@
+package ring
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	totalTokensCount         = 1 << 32
+	optimalTokensPerInstance = 1 << 9
+	maxZonesCount            = 1 << 3
+)
+
+var (
+	instanceIDRegex          = regexp.MustCompile(`^(.*)-(zone-[a-z])-(\d+)$`)
+	errorBadInstanceIDFormat = func(instanceID string) error {
+		return fmt.Errorf("unable to extract instance id from \"%s\"", instanceID)
+	}
+	errorZoneNotValid = func(zone string) error {
+		return fmt.Errorf("zone %s is not valid", zone)
+	}
+	errorMultipleOfZonesCount = func(optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: optimal token ownership must be a positive multiple of maximal allowed number of zones %d", token.prevToken, token.token, optimalTokenOwnership, maxZonesCount)
+	}
+	errorLowerAndUpperBoundModulo = func(optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: lower and upper bounds must be congruent modulo maximal allowed number of zones %d", token.prevToken, token.token, optimalTokenOwnership, maxZonesCount)
+	}
+	errorDistanceBetweenTokensNotBigEnough = func(optimalTokenOwnership, ownership int, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: distance between lower and upper bound %d is not big enough", token.prevToken, token.token, optimalTokenOwnership, ownership)
+	}
+)
+
+type SpreadMinimizingConfig struct {
+	InstanceID string
+	Zone       string
+}
+
+type SpreadMinimizingTokenGenerator struct {
+	cfg        SpreadMinimizingConfig
+	instanceID int
+	zoneID     int
+	zones      []string
+	logger     log.Logger
+}
+
+func NewSpreadMinimizingTokenGenerator(cfg SpreadMinimizingConfig, zones []string, logger log.Logger) (*SpreadMinimizingTokenGenerator, error) {
+	sortedZones := make([]string, len(zones))
+	copy(sortedZones, zones)
+	if !slices.IsSorted(sortedZones) {
+		sort.Strings(sortedZones)
+	}
+	instanceID, err := parseInstanceID(cfg.InstanceID)
+	if err != nil {
+		return nil, err
+	}
+	zoneID, err := findZoneID(cfg.Zone, sortedZones)
```

Comment:
I'm a bit confused by the concept of zone ID here, because I don't know which actual value will be passed as `zones`. Let's say you're creating a new cluster and the first instance registers in the ring, and it's from `zone-b`. What will be the value of `zones` passed to `NewSpreadMinimizingTokenGenerator()`?

Reply:
I need an identifier that will distinguish different zones. 
If we have zones `zone-a`, `zone-b` and `zone-c`, I would need to assign them numbers `0`,`1` and `2` (in any order, but the same zone must always be assigned to the same number), because in the approach I am introducing tokens from the same zone must be equal modulo `maxZoneCount` to each other.

Reply:
Ok, so in practice we expect input `zones` to always be `[]string{"zone-a", "zone-b", "zone-c"}` even when the 1st instance is registered to the ring and the ring is empty.